src/Nirum/Parser.hs

```diff
@@ -627,28 +629,45 @@ modulePath = do
     f Nothing i = Just $ ModuleName i
     f (Just p) i = Just $ ModulePath p i
 
-importName :: Parser (Identifier, A.AnnotationSet)
-importName = do
+importName :: [Identifier]
+           -> Parser (Identifier, Identifier, A.AnnotationSet)
+importName forwardNames = do
     aSet <- annotationSet <?> "import annotations"
     spaces
     iName <- identifier <?> "name to import"
-    return (iName, aSet)
-
-imports :: Parser [TypeDeclaration]
-imports = do
+    aName <- optional $ try $ do
+      spaces
+      string' "as"
+      spaces
+      n <- uniqueIdentifier forwardNames "alias name to import"
+      spaces
+      return n
+    return (iName, fromMaybe iName aName, aSet)
+
+imports :: [Identifier] -> Parser [TypeDeclaration]
+imports forwardNames = do
     string' "import" <?> "import keyword"
     spaces
     path <- modulePath <?> "module path"
     spaces
     char '('
     spaces
-    idents <- (importName >>= \ i -> spaces >> return i)
-        `sepEndBy1` (char ',' >> spaces)
-        <?> "names to import"
+    idents <- many' [] $ \ importNames' -> try $ do
+        let forwardNames' = [ i | (_, i, _) <- importNames' ] ++
+                forwardNames
+        unless (L.null importNames') $ do
+            string' ","
+            spaces
+        importName forwardNames'
+    when (L.null idents) $ fail "parentheses cannot be empty"
+    void $ optional $ string' ","
+    spaces
     char ')'
     spaces
     char ';'
-    return [Import path ident aSet | (ident, aSet) <- idents]
+    return [ Import path source imp aSet
+           | (imp, source, aSet) <- idents
+           ]
```

Comment:
We should have tests for all combinations of the following conditions:

|    | Empty parentheses | Single import name w/o trailing comma | Single import name w/ trailing comma | Multiple import names w/o trailing comma | Multiple import names w/ trailing comma |
| --- | --- | --- | --- | --- | --- |
| With `as` | `import foo ()` | `import foo (x as a)` | `import foo (x as a,)` | `import foo (x as a, y as b, c)` | `import foo (x as a, y as b, c,)` |
| Without `as` | `import foo()` | `import foo (a)` | `import foo(a, b, c,)` | `import foo (a,)` | `import foo(a, b, c)` |

Reply:
We should test if the parser disallows and shows an error when there are duplicated alias names.